projection/validator_delegation/handle_msg.go

```diff
@@ -0,0 +1,329 @@
+	"github.com/crypto-com/chain-indexing/projection/validator_delegation/view"
+	"github.com/crypto-com/chain-indexing/usecase/coin"
+)
+
+func (projection *ValidatorDelegation) handleCreateNewValidator(
+	rdbTxHandle *rdb.Handle,
+	height int64,
+	validatorAddress string,
+	delegatorAddress string,
+	tendermintPubKey string,
+	amount coin.Int,
+	minSelfDelegationInString string,
+) error {
+
+	consensusNodeAddress, err := utils.GetConsensusNodeAddress(tendermintPubKey, projection.config.conNodeAddressPrefix)
+	if err != nil {
+		return fmt.Errorf("error in GetConsensusNodeAddress: %v", err)
+	}
+
+	tendermintAddress, err := utils.GetTendermintAddress(tendermintPubKey)
+	if err != nil {
+		return fmt.Errorf("error in GetTendermintAddress: %v", err)
+	}
+
+	minSelfDelegation, ok := coin.NewIntFromString(minSelfDelegationInString)
+	if !ok {
+		return fmt.Errorf("Failed to parse minSelfDelegationInString: %v", minSelfDelegationInString)
+	}
+
+	// Insert an ValidatorRow.
+	validatorsView := NewValidators(rdbTxHandle)
+	validatorRow := view.ValidatorRow{
+		Height: height,
+
+		OperatorAddress:      validatorAddress,
+		ConsensusNodeAddress: consensusNodeAddress,
+		TendermintAddress:    tendermintAddress,
+
+		Status: types.UNBONDED,
+		Jailed: false,
+		Power:  "0",
+
+		UnbondingHeight: int64(0),
+		UnbondingTime:   utctime.UTCTime{},
+
+		Tokens:            coin.ZeroInt(),
+		Shares:            coin.ZeroDec(),
+		MinSelfDelegation: minSelfDelegation,
+	}
+	if err := validatorsView.Insert(validatorRow); err != nil {
+		return fmt.Errorf("error validatorsView.Insert(): %v", err)
+	}
+
+	if _, err := projection.handleDelegate(
+		rdbTxHandle,
+		height,
+		validatorAddress,
+		delegatorAddress,
+		amount,
+	); err != nil {
+		return fmt.Errorf("error in projection.handleDelegate(): %v", err)
+	}
+
+	return nil
+}
+
+func (projection *ValidatorDelegation) handleEditValidator(
+	rdbTxHandle *rdb.Handle,
+	height int64,
+	validatorAddress string,
+	minSelfDelegationInString string,
+) error {
+
+	validatorsView := NewValidators(rdbTxHandle)
+
+	// Get the validator
+	validator, found, err := validatorsView.FindByOperatorAddr(validatorAddress, height)
+	if err != nil {
+		return fmt.Errorf("error validatorsView.FindByOperatorAddr(): %v", err)
+	}
+	if !found {
+		return fmt.Errorf("error validator not found: %v, %v", validatorAddress, height)
+	}
+
+	minSelfDelegation, ok := coin.NewIntFromString(minSelfDelegationInString)
+	if !ok {
+		return fmt.Errorf("Failed to parse minSelfDelegationInString: %v", minSelfDelegationInString)
+	}
+	// Update Valdiator
+	validator.MinSelfDelegation = minSelfDelegation
+
+	// Write the updated Validator to DB
+	if err := validatorsView.Update(validator); err != nil {
+		return fmt.Errorf("error validatorsView.Update(): %v", err)
+	}
+
+	return nil
+}
+
+// handleDelegate - Handle create/update delegation. Return newShares and error.
+func (projection *ValidatorDelegation) handleDelegate(
```

Comment:
are this handler bond to `MsgDelegate` ? if yes, could we name it handleMsgDelegate?

Reply:
Not really, when handling `MsgCreateValidator`, it will also be invoked.